modules/node-shared/src/main/scala/io/constellationnetwork/node/shared/infrastructure/consensus/ValidationErrorStorage.scala

```diff
@@ -0,0 +1,54 @@
+package io.constellationnetwork.node.shared.infrastructure.consensus
+
+import java.time.Instant
+
+import cats.data.NonEmptyList
+import cats.effect.kernel.{Async, Clock, Ref}
+import cats.effect.std.Semaphore
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+
+trait ValidationErrorStorage[F[_], Artifact, ValidationError] {
+  def addValidationError(artifact: Artifact, errors: ValidationError): F[Unit]
+  def addValidationErrors(errors: NonEmptyList[(Artifact, ValidationError)]): F[Unit]
+  def allValidationErrors: F[Vector[(Artifact, ValidationError, Instant)]]
+}
+
+object ValidationErrorStorage {
+
+  def make[F[_]: Async, Artifact, ValidationError](
+    maxSize: Int
+  )(implicit clock: Clock[F]): F[ValidationErrorStorage[F, Artifact, ValidationError]] =
+    for {
+      updateSemaphore <- Semaphore[F](1)
+      validationErrorsR <- Ref[F].of(Vector.empty[(Artifact, ValidationError, Instant)])
+    } yield
+      new ValidationErrorStorage[F, Artifact, ValidationError] {
+        override def addValidationError(artifact: Artifact, errors: ValidationError): F[Unit] =
+          updateSemaphore.permit.use { _ =>
+            addError(artifact, errors)
+          }
+
+        override def allValidationErrors: F[Vector[(Artifact, ValidationError, Instant)]] = validationErrorsR.get
+
+        override def addValidationErrors(errors: NonEmptyList[(Artifact, ValidationError)]): F[Unit] =
+          updateSemaphore.permit.use { _ =>
+            errors.foldLeft(addError(errors.head._1, errors.head._2)) { (acc, x) =>
+              for {
+                _ <- acc
+                res <- addError(x._1, x._2)
```

Comment:
btw. you don't need a for comprehension for this. 

instead of
```scala
for {
  _ <- acc
  res <- addError(x._1, x._2)
} yield res
```

you could write:
```scala
acc >> addError(x._1, x._2)
```